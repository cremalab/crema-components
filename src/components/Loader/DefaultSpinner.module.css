/* from: https://loading.io/css/ */

.DefaultSpinner {
  display: inline-block;
  position: relative;
  width: 40px;
  height: 40px;
}

.DefaultSpinner.small {
  width: 20px;
  height: 20px;
}

.DefaultSpinner.medium {
  width: 40px;
  height: 40px;
}

.DefaultSpinner.large {
  width: 60px;
  height: 60px;
}

.DefaultSpinner div {
  box-sizing: border-box;
  display: block;
  position: absolute;
  width: 100%;
  height: 100%;
  margin: 8px;
  border: 8px solid #000;
  border-radius: 50%;
  animation: DefaultSpinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
  border-color: #000 transparent transparent transparent;
}

.DefaultSpinner.small div {
  border-width: 2px;
}

.DefaultSpinner.medium div {
  border-width: 4px;
}

.DefaultSpinner.large div {
  border-width: 6px;
}

.DefaultSpinner div:nth-child(1) {
  animation-delay: -0.45s;
}

.DefaultSpinner div:nth-child(2) {
  animation-delay: -0.3s;
}

.DefaultSpinner div:nth-child(3) {
  animation-delay: -0.15s;
}

@keyframes DefaultSpinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
