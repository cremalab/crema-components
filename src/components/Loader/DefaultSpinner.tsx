import styles from "./DefaultSpinner.module.css"

interface DefaultSpinnerProps {
  size?: "small" | "medium" | "large"
}

export function DefaultSpinner({ size = "medium" }: DefaultSpinnerProps) {
  return (
    <div
      className={`${styles.DefaultSpinner} ${styles[size]}`}
      aria-hidden="true"
    >
      <div></div>
      <div></div>
      <div></div>
      <div></div>
    </div>
  )
}
