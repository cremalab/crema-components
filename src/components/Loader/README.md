# `<Loader />`

This component is intended to be used a general purpose loading indicator throughout the application. The main custom aspect is that the Loader has a `timeout` in which it will wait that amount of time (in milliseconds) before actually rendering the Loader, before that it renders `null`. This is intended to reduce the amount of "jank" caused by a loading indicator showing up for less than 200 milliseconds before whatever it was waiting on loads in.

By default the Loader uses a custom "spinner" to indicate loading status, but that can be swapped out using `children`.

## Directory Structure

- `Loader.stories.tsx`: Component stories (`npm run test:playground`)
- `Loader.test.tsx`: Component tests (`npm run test:unit`)
- `Loader.tsx`: Component code
- `index.ts`: Component export
- `README.md`: Component documentation (hey, that's me!)
