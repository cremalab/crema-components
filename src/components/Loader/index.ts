export { Loader } from "./Loader"
export { DefaultSpinner } from "./DefaultSpinner"
