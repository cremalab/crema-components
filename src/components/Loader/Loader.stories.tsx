import { ComponentMeta, Story } from "@storybook/react"
import { ComponentProps } from "react"
import { Loader } from "./Loader"

/**
 * See Storybook Docs: Writing Stories
 * https://storybook.js.org/docs/basics/writing-stories/
 */

export default {
  title: "Components/Loader",
} as ComponentMeta<typeof Loader>

const Template: Story<ComponentProps<typeof Loader>> = (args) => (
  <Loader {...args} />
)

export const Default = Template.bind({})

Default.args = {
  timeout: 500,
}

export const WithChildren = Template.bind({})

WithChildren.args = {
  timeout: 500,
  children: <div>Loading</div>,
}
