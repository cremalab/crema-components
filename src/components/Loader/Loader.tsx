import { ReactNode, useEffect, useState } from "react"
import { DefaultSpinner } from "./DefaultSpinner"

export interface LoaderProps {
  ariaLabel?: string
  timeout?: number
  children?: ReactNode
}

export function Loader({
  // this is the text that will be read to the screen reader since most loaders are visual in nature
  ariaLabel = "Loading now please wait",
  timeout = 200,
  children = <DefaultSpinner />,
}: LoaderProps) {
  // If the timeout is 0 we short circuit the logic and immediately show the loader to reduce re-renders
  const [showLoader, setShowLoader] = useState(timeout === 0)

  useEffect(() => {
    let timeoutId: NodeJS.Timeout | null = null

    // if the timeout is not 0 we need to call setTimeout with that value to re-render and show the loader
    if (timeout !== 0) {
      timeoutId = setTimeout(() => {
        setShowLoader(true)
      }, timeout)
    }

    return () => {
      if (timeoutId) {
        // always make sure to clean up timeouts :)
        clearTimeout(timeoutId)
      }
    }
  }, [timeout])

  if (!showLoader) {
    return null
  }

  // these aria attributes should inform the screen reader that this is a progress indicator that is likely to be changing
  return (
    <div
      role="progressbar"
      aria-busy="true"
      aria-live="polite"
      aria-label={ariaLabel}
    >
      {children}
    </div>
  )
}
