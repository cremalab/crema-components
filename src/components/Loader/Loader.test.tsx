import { act, render, screen } from "@testing-library/react"
import { DefaultSpinner } from "./DefaultSpinner"
import { Loader } from "./Loader"

describe("Loader", () => {
  it("renders nothing before the timeout", () => {
    // Arrange
    const text = "Loading"
    const timeout = 1000

    jest.useFakeTimers()

    // Act
    render(<Loader timeout={timeout}>{text}</Loader>)

    act(() => {
      jest.advanceTimersByTime(timeout / 2)
    })

    const received = screen.queryByText(text)

    // Assert
    expect(received).toBeNull()
  })

  it("renders the expected content after the timeout", () => {
    // Arrange
    const text = "Loading"
    const ariaLabel = "Now loading please wait"
    const timeout = 1000

    jest.useFakeTimers()

    // Act
    render(
      <Loader timeout={timeout} ariaLabel={ariaLabel}>
        {text}
      </Loader>,
    )

    act(() => {
      jest.advanceTimersByTime(timeout)
    })

    const received = screen.queryByLabelText(ariaLabel)

    // Assert
    expect(received).toBeInTheDocument()
    expect(received).toHaveTextContent(text)
  })

  it("renders instantly if the timeout is 0", () => {
    // Arrange
    const text = "Loading"
    const ariaLabel = "Now loading please wait"
    const timeout = 0

    jest.useFakeTimers()

    // Act
    render(
      <Loader timeout={timeout} ariaLabel={ariaLabel}>
        {text}
      </Loader>,
    )

    const received = screen.queryByLabelText(ariaLabel)

    // Assert
    expect(received).toBeInTheDocument()
    expect(received).toHaveTextContent(text)
  })
})

describe("DefaultSpinner", () => {
  it("renders the expected structure with size = small", () => {
    // Arrange
    const size = "small"

    // Act
    const { asFragment } = render(<DefaultSpinner size={size} />)

    // Assert
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders the expected structure with size = medium", () => {
    // Arrange
    const size = "medium"

    // Act
    const { asFragment } = render(<DefaultSpinner size={size} />)

    // Assert
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders the expected structure with size = large", () => {
    // Arrange
    const size = "large"

    // Act
    const { asFragment } = render(<DefaultSpinner size={size} />)

    // Assert
    expect(asFragment()).toMatchSnapshot()
  })
})
