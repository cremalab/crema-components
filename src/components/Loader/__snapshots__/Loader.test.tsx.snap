// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultSpinner renders the expected structure with size = large 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="DefaultSpinner large"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</DocumentFragment>
`;

exports[`DefaultSpinner renders the expected structure with size = medium 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="DefaultSpinner medium"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</DocumentFragment>
`;

exports[`DefaultSpinner renders the expected structure with size = small 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="DefaultSpinner small"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</DocumentFragment>
`;
